### PR TITLE
add `with_doi` method in query builder

### LINKED ISSUE #140: As a LLM user, I want to have an idea of the number of result returned
### Checked for duplicates

No - I haven't checked

### 🧑‍🔬 User Persona(s)

User of the MCP server.

### 💪 Motivation

...so that I can adjust intuitively the granularity I expect in my results, e.g. switch from datasets to collections when I have too many results.

Maybe the default 50 limit is not what we should keep used by the MCP client is not what we should keep (as tested with claude code).

### 📖 Additional Details

By returning the total number of results in the peppi Products object. With a `count` property which would trigger the first lazy call to the API.

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### 🎉 I&T

_No response_

### LINKED ISSUE #158: As a data user, I want to search for products by DOI
## Checked for duplicates
Yes - I've already checked

## 🧑‍🔬 User Persona(s)
Data User, Researcher, Data Librarian, Publication Author

## 💪 Motivation
...so that I can quickly find and access specific datasets referenced in papers or documentation, improving data traceability and reproducibility in research.

## 📖 Additional Details

**Current Behavior:**
- ✅ DOI information is available in product metadata as `pds:Citation_Information.pds:doi`
- ✅ Users can retrieve DOI from products after finding them
- ❌ The `get()` method only supports LIDVID identifiers, not DOIs
- ❌ No direct way to search BY DOI

**Proposed Solution:**
Enhance the `get()` method to auto-detect identifier type (LID/LIDVID/DOI):

```python
# Should work with LID
products.get('urn:nasa:pds:context:target:planet.mars')

# Should work with LIDVID  
products.get('urn:nasa:pds:context:target:planet.mars::1.0')

# Should work with DOI (NEW)
products.get('10.1234/example-doi')

# Alternative: Dedicated method
products.has_doi('10.1234/example-doi')
```

**Implementation Approach:**
1. Extend `get()` to accept `Union[LID, LIDVID, DOI]` as string
2. Auto-detect identifier type:
   - Starts with "urn:" → LID/LIDVID
   - Starts with "10." → DOI
   - Pattern match for DOI format
3. Query appropriate field based on identifier type
4. Consider dedicated `has_doi()` method for explicitness

**Use Case:**
When reading a paper that cites "DOI: 10.1234/pds-example", users can directly search:
```python
product = pep.Products(client).get('10.1234/pds-example')
```

This is the inverse of the existing workflow (Cookbook recipe #9 shows retrieving DOIs from found products).

**Related:**
- Part of Priority Requirements PDF (Requirement 7)
- Similar to #29 (target LID), #30 (investigation LID), #31 (instrument LID)
- Replaces closed issue NASA-PDS/peppi#143

---
## Acceptance Criteria
**Given** a user has a DOI from a publication
**When I perform** `products.get('10.1234/example-doi')`
**Then I expect** the product(s) matching that DOI

## ⚙️ Engineering Details
_To be filled by Engineering Node Team_

## 🎉 I&T
_To be filled by Engineering Node Team_

### PR DESCRIPTION
**Pull request by tloubrieu-jpl**
## 🗒️ Summary

While preparing the training for the Management Council, as an example, I added the search by DOI method in peppi and later the count() method.

### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [x] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [ ] AI generated substantial portions of this code

Estimated % of code influenced by AI: ___ %

## ⚙️ Test Data and/or Report

New unit tests added.
count test includes test of side effects with  the pagination of the actual results.

Automated non regression successful.

## ♻️ Related Issues

- fixes https://github.com/NASA-PDS/peppi/issues/158
- fixes https://github.com/NASA-PDS/peppi/issues/140

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

**Review by nutjob4life — APPROVED** [review]
Code delta: ✓
Tox: ✓ (took 11m!)
Approval: ✅
Tox details:
```
  py312: OK (329.91=setup[0.77]+cmd[329.14] seconds)
  py313: OK (326.09=setup[1.45]+cmd[324.64] seconds)
  docs: OK (3.69=setup[0.69]+cmd[3.00] seconds)
  lint: OK (2.09=setup[0.01]+cmd[2.09] seconds)
  congratulations :) (661.89 seconds)
```

File: src/pds/peppi/query_builder.py [hunk 1]
```diff
@@ -515,3 +550,22 @@ def reset(self):
...
+    def count(self) -> int:
+        """Returns the number of products returned by this query.
+
+        This requires fetching the first page of results from the PDS Registry API to get the total hit count, but does not require iterating through all results to get a count. If the count has already been fetched during a previous iteration over results, this method will return the cached count value without making an additional API request.
+
+        """
+
+        if self._result_set._count is None:
+            # if not done yet, init a new page to get the count
+            self._result_set.init_new_page(query_string=self._q_string, fields=self._fields)
```
[the remark this answers is not in the document]
Comment (jordanpadams) [reply src/pds/peppi/query_builder.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/NASA-PDS/peppi/pull/159#discussion_r2907797758)
